File: Data Structures & Algorithms/climbing-stairs/submission-8.java
class Solution {
    public int climbStairs(int n) {
        int[] dp = new int[n+1];
        dp[0] = 1;
        dp[1] = 1;
        for(int i=2; i<= n; i++){
            dp[i] = dp[i-1] + dp[i-2];
        }
        
        return dp[n];
    }

    // private int dfs(int n, int[] memo){
    //     if(memo[n] != -1){
    //         return memo[n];
    //     }

    //     if(n == 0 || n == 1){
    //         return 1;
    //     }

    //     memo[n] = dfs(n-1, memo) + dfs(n-2, memo);
    //     return memo[n];
    // }
}

File: Data Structures & Algorithms/word-break/submission-8.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        n = len(s)
        memo = {}

        def dfs(i):
            if i >= n:
                return True

            if i in memo:
                return memo[i]

            for w in wordDict:
                if i + len(w) <= n and w == s[i:i+len(w)]:
                    if dfs(i+len(w)):
                        memo[i] = True
                        return True
                        
            memo[i] = False
            return False

        return dfs(0)
                    
